subsys/net/lib/tls_credentials/tls_credentials_shell.c

```diff
@@ -0,0 +1,813 @@
+		*out = CRED_STORAGE_FMT_STRING;
+		*terminated = true;
+		return 0;
+	}
+
+	shell_fprintf(
+		sh, SHELL_ERROR,
+		"%s is not a valid storage format.\n", arg
+	);
+
+	return -EINVAL;
+}
+
+/* Clear credential buffer, with shell feedback */
+static void shell_clear_cred_buf(const struct shell *sh)
+{
+	/* We will only print a message if some data was actually wiped. */
+	if (cred_buf_clear()) {
+		shell_fprintf(
+			sh, SHELL_NORMAL,
+			"Credential buffer cleared.\n"
+		);
+	}
+}
+
+/* Write data into the credential buffer, with shell feedback. */
+static int shell_write_cred_buf(const struct shell *sh, char *chunk)
+{
+	int res = cred_buf_write(chunk);
+
+	/* Report results. */
+
+	if (res == -ENOMEM) {
+		shell_fprintf(
+			sh, SHELL_ERROR,
+			"Not enough room in credential buffer for provided data. "
+			"Increase CONFIG_TLS_CREDENTIALS_SHELL_CRED_BUF_SIZE.\n"
+		);
+		shell_clear_cred_buf(sh);
+		return -ENOMEM;
+	}
+
+	shell_fprintf(
+		sh, SHELL_NORMAL,
+		"Stored %d bytes.\n", res
+	);
+
+	return 0;
+}
+
+/* Adds a credential to the credential store */
+static int tls_cred_cmd_add(const struct shell *sh, size_t argc, char *argv[])
+{
+	int err = 0;
+	sec_tag_t sectag;
+	enum cred_storage_fmt format;
+	bool terminated;
+	enum tls_credential_type type;
+	void *cred_copy = NULL;
+	void *cred_chosen = NULL;
+	bool keep_copy = false;
+
+	/* Lock credentials so that we can interact with them directly.
+	 * Mainly this is required by credential_get.
+	 */
+
+	credentials_lock();
+
+	err = shell_parse_cred_sectag(sh, argv[1], &sectag, false);
+	if (err) {
+		goto cleanup;
+	}
+
+	err = shell_parse_cred_type(sh, argv[2], &type, false);
+	if (err) {
+		goto cleanup;
+	}
+
+	err = shell_parse_cred_backend(sh, argv[3]);
+	if (err) {
+		goto cleanup;
+	}
+
+	err = shell_parse_cred_storage_format(sh, argv[4], &format, &terminated);
+	if (err) {
+		goto cleanup;
+	}
+
+	if (argc == 6) {
+		/* Credential was passed, clear credential buffer and then use the passed-in
+		 * credential.
+		 */
+		shell_clear_cred_buf(sh);
+		err = shell_write_cred_buf(sh, argv[5]);
+
+		if (err) {
+			goto cleanup;
+		}
+	}
+
+	/* Make sure the credential buffer isn't empty. */
+	if (cred_written == 0) {
+		shell_fprintf(
+			sh, SHELL_ERROR,
+			"Please provide a credential to add.\n"
+		);
+		err = -ENOENT;
+		goto cleanup;
+	}
+
+	/* Check whether a credential of this type and sectag already exists. */
+	if (credential_get(sectag, type)) {
+		shell_fprintf(
+			sh, SHELL_ERROR,
+			"TLS credential with sectag %d and type %s already exists.\n", sectag,
+			cred_type_name(type)
+		);
+		err = -EEXIST;
+		goto cleanup;
+	}
+
+	/* If binary format was specified, decode from base64. */
+	if (format == CRED_STORAGE_FMT_BINARY) {
+		/* base64_decode can handle in-place operation.
+		 * Pass &cred_written as olen so that it is updated to match the size of the base64
+		 * encoding.
+		 *
+		 * We use sizeof(cred_buf)-1 since we want to keep room for a NULL terminator.
```

Comment:
I think this needs some elaboration, I don't follow the idea. What is the point of NULL termination of  binary data? 0 (NULL) could be present anywhere within the data itself. Same comment goes for the idea of `bint` data format.

Reply:
My thinking was:

- Some TLS credential types and formats require NULL termination, some do not
- Some credentials may be enter-able as plaintext, some credentials may not
- These two requirements are not tightly coupled

So I may as well just let the user determine each individually.

To be more specific, consider the following motivating cases:

If the credential is a PSK, it may technically be pure binary, in which case the `bin` format is the correct choice, since PSK do not need to be NULL-terminated.

But, if the PSK happens to be a plain-text/user-enterable PSK, then for convenience, the `str` format is provided. Again, without NULL-termination since that is the expectation for PSKs.

Meanwhile, X509 certs can be provided to the TLS Credentials library either as a raw-binary buffer, or as a PEM-formatted string.

If the user opts for just using raw-binary, they may simply provide the base-64 encoding of that binary, and use the `bin` format.

The PEM format, on the other hand, is expected to be provided as a typical C-string, so ideally the user would be able to use the `strt` format. But PEM also requires newlines, which the Zephyr shell does not support, so `strt` is off the table.

That means the user is required to use either `bin` or `bint` when providing PEM-encoded certs, and must provide them by first base-64 encoding the (already base-64 encoded) PEM string.

Then, the difference between `bint` and `bin` is whether or not the user also has to manually encode their own NULL terminator.

Given all of the above, you are right, we could technically get away with just `bin` and `str` (the **non**-NULL-terminated options).

But for the sake of intuition, I thought it would be wise to support `strt`, since that's what's required to produce a traditional C-string from human-readable text input.

And once I added `bin`, `str`, and `strt`, adding `bint` seemed only natural.

In truth, though, I'm not 100% sold on the above. I'd be interested to hear your thoughts. Do you think just supporting `bin` and `str` would be better? Is `str` worth supporting at all when users could just use bin with the base64 encoding of whatever strings they want?

Reply:
Maybe it would be better to drop support for `bint`, but keep `str` and `strt`? Realistically, I doubt anyone will actually need to enter credentials in PEM format via the shell -- they can instead just copy the base64 portion of the PEM and use `bin` mode. Which means that the use-case for `bint` is less important.

Besides, `bint` only makes sense because Zephyr shell doesn't support multi-line string entry. If support for that is ever added, suddenly the `bint` format won't make sense. So maybe it is better to remove it now and sidestep that being a potential problem.

I'd be on board with going this route assuming @rlubos and @plskeggs you both agree

Reply:
Thanks for clarifying of the use case of `bint`.

I was leaning towards the idea of dropping `bint` support in favour of base64-encoded DER certs, but as you now highlight that there is an actual use-case for it (cert chains) perhaps we need to keep it. Maybe just document it better, in what cases should this format be used (I can only think of this PEM case)?

Reply:
Alright, great. Yeah, some more explanation could be useful. I've gone ahead and added that just after the table of format specifiers. Let me know what you think

Reply:
Looks good, thanks.